pinc/upload_file.inc

```diff
@@ -0,0 +1,255 @@
+
+    // Display a div with a form; some content which can be abritrary HTML/other inputs;
+    // a 'choose file' button and a labeled submit button
+    echo "<div style='border: 1px solid grey; margin-left: .5em; padding: .25em;'>\n";
+
+    echo "<form id='upload_form' method='POST' enctype='multipart/form-data'>\n";
+    echo "<input type='hidden' name='mode' value='upload'>";
+    echo $form_content;
+
+    echo "<p>$standard_blurb</p>";
+    // these are the form elements for non-resumable upload
+    echo "<div id='old_uploader'>
+    <p class='warning'>", sprintf(_('Maximum file size is %s.'), get_max_upload_size()), "</p>
+    <p>", get_big_upload_blurb(), "</p>",
+    _("File to upload"), ": <input type='file' id='old_browse' accept='.zip' name='uploaded_file' size='25' maxsize='50'>
+    <input type='submit' id='old_submit' value='", attr_safe($submit_label), "'>",
+    "<p>", sprintf(_("After you click the '%s' button, the browser will appear to be slow getting to the next page. This is because it is uploading the file."), $submit_label), "</p>",
+    "</div>";
+
+    // if resumable, these are shown instead
+    echo "<div id='resumable_uploader' style='display: none;'>
+    <input type='hidden' name='resumable_filename' value=''>
+    <input type='hidden' name='resumable_identifier' value=''>
+    <p>", sprintf(_("Your browser supports uploading large files (up to %s) via javascript. If the upload fails,
+        or you navigate away from this page before it finishes, uploading the file again will pick up where it left off."), humanize_bytes(RESUMABLE_UPLOAD_SIZE)), "</p>
+    <p>", _("File to upload"), ": <span id='resumable_selected_file'></span> &nbsp;
+    <span id='resumable_browse' class='button'>", _("Choose File"), "</span>
+    <span id='resumable_submit' class='button'>", _("Upload"), "</span></p>
+    </div>";
+
+    echo "<p>", _("Upload progress"), ":&nbsp;<span id='upload_progress'></span></p>";
+
+    echo "</form>\n";
+    echo "</div>\n";
+}
+
+// return null if there is no file
+// or return a file_info with 'name' and 'tmp_name'
+// throw a FileException if there is a probem
+function validate_uploaded_file()
+{
+    $file_path = "";
+    try
+    {
+        $mode = get_enumerated_param($_REQUEST, 'mode', 'upload', ['upload', 'resumable']);
+        if('upload' == $mode)
+        {
+            $file_info = $_FILES['uploaded_file'];
+
+            // If a user hits the "Upload" button without first selecting a file,
+            // it appears that most browsers send a request containing a file whose
+            // name and content are empty. But I think it's also legal for a browser
+            // to send a request that doesn't contain a file at all (in which case
+            // $file_info would be null.  Check both possibilities.
+            if (is_null($file_info) || $file_info['name'] == '') {
+                return NULL;
```

Comment:
Is there a reason we shouldn't just throw an exception here anyway? That would streamline how this function is used -- it either returns a valid $file_info or it raises an exception.

Reply:
You could even raise and throw separate exception when no files was uploaded, that way you could catch and handle that expected case in `upload_text.php`.